### PR TITLE
Refactor `BitOps` to not rely on `__CUDA_ARCH__` macro

### PR DESCRIPTION
**Pull request by dalg24**
Following up on #4660

### CODE REVIEW

**Comment by dalg24** [comment]
Retest this please

File: core/src/impl/Kokkos_BitOps.hpp [hunk 1]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
+  if (i) {
+    for (offset = 0; (i & (1 << offset)) == 0; ++offset)
       ;
   }
   return offset;
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION inline int bit_scan_forward_device(unsigned i) {
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
+  return __popc(i);
```
Comment (crtrott) [inline core/src/impl/Kokkos_BitOps.hpp:172]: why popc and not fffs(i) -1 ?

**Review by crtrott — CHANGES_REQUESTED** [review]
I think there is one mistake for bit_scan_forward

File: core/src/impl/Kokkos_BitOps.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/impl/Kokkos_BitOps.hpp:172]: Copy/paste accident...

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

File: core/src/impl/Kokkos_BitOps.hpp [hunk 2]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
 namespace Impl {
 
 KOKKOS_FORCEINLINE_FUNCTION
-int int_log2(unsigned i) {
+int int_log2_fallback(unsigned i) {
   enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:61]:
```suggestion
  constexpr int shift = sizeof(unsigned) * CHAR_BIT - 1;
```

File: core/src/impl/Kokkos_BitOps.hpp [hunk 3]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
+  if (i) {
+    for (offset = shift; (i & (1 << offset)) == 0; --offset)
+      ;
+  }
+  return offset;
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int int_log2_device(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:72]:
```suggestion
  constexpr int shift = sizeof(unsigned) * CHAR_BIT - 1;
```

File: core/src/impl/Kokkos_BitOps.hpp [hunk 4]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
   return shift - __clz(i);
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return int_log2_fallback(i);
+#endif
+}
+
+KOKKOS_IMPL_HOST_FUNCTION
+inline int int_log2_host(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:82]:
```suggestion
  constexpr int shift = sizeof(unsigned) * CHAR_BIT - 1;
```

File: core/src/impl/Kokkos_BitOps.hpp [hunk 5]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int int_log2_device(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return shift - __clz(i);
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return int_log2_fallback(i);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:76]: I believe
```suggestion
#elif defined(KOKKOS_COMPILER_INTEL)
  return _bit_scan_reverse(i);  
#else
  return int_log2_fallback(i);
```
would be needed to not change behavior for for Intel compilers on the device.

File: core/src/impl/Kokkos_BitOps.hpp [hunk 6]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
 
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int bit_first_zero_device(unsigned i) noexcept {
+  enum : unsigned { full = ~0u };
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return full != i ? __ffs(~i) - 1 : -1;
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return bit_first_zero_fallback(i);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:128]:
```suggestion
#elif defined(KOKKOS_COMPILER_INTEL)
  return full != i ? _bit_scan_forward(~i) : -1;
#else
  return bit_first_zero_fallback(i);
```
... and here as well.

File: core/src/impl/Kokkos_BitOps.hpp [hunk 7]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
 
-KOKKOS_FORCEINLINE_FUNCTION
-int bit_scan_forward(unsigned i) {
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+KOKKOS_IMPL_DEVICE_FUNCTION inline int bit_scan_forward_device(unsigned i) {
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return __ffs(i) - 1;
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return bit_scan_forward_fallback(i);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:174]: ... and here
```suggestion
#elif defined(KOKKOS_COMPILER_INTEL)
  return _bit_scan_forward(i);
#else
  return bit_scan_forward_fallback(i);
```

File: core/src/impl/Kokkos_BitOps.hpp [hunk 8]
```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
...
+  // count
+  return (int)((i * (~0u / 255u)) >> (sizeof(unsigned) - 1) * CHAR_BIT);
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION inline int bit_count_device(unsigned i) {
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return __popc(i);
-#elif defined(__INTEL_COMPILER)
+#else
+  return bit_count_fallback(i);
```
Comment (masterleinad) [inline core/src/impl/Kokkos_BitOps.hpp:219]: ... and here
```suggestion
#elif defined(__INTEL_COMPILER)
  return _popcnt32(i);
#else
  return bit_count_fallback(i);
```

File: core/src/impl/Kokkos_BitOps.hpp (hunk 5 above)
Comment (dalg24) [reply core/src/impl/Kokkos_BitOps.hpp:77]: Done

File: core/src/impl/Kokkos_BitOps.hpp [hunk 9]
```diff
@@ -57,93 +57,202 @@ namespace Kokkos {
...
   return __popc(i);
 #elif defined(__INTEL_COMPILER)
   return _popcnt32(i);
+#else
+  return bit_count_fallback(i);
+#endif
+}
+
+KOKKOS_IMPL_HOST_FUNCTION inline int bit_count_host(unsigned i) {
+#if defined(__INTEL_COMPILER)
```
Comment (dalg24) [inline core/src/impl/Kokkos_BitOps.hpp:233]: Note this is using the `__INTEL_COMPILER` macro directly

**Comment by dalg24** [comment]
Retest this please

File: core/src/impl/Kokkos_BitOps.hpp (hunk 9 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_BitOps.hpp:236]: I can confirm that `icpx` is fine with `_popcnt32` even on the device so using `KOKKOS_COMPILER_INTEL`  should be fine here (and then also using on the device).
I'm happy to create a follow-up pull request, if you don't want to change it here.

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Comment by masterleinad** [comment]
Retest this please.

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring HPX failure and not waiting for SYCL build

**Pull request merged**